### PR TITLE
Add xfs offline transport

### PR DESCRIPTION
**Pull request by Felixoid**
- Add xfsdump/xfsrestore offline migration transport
- Add disk changing during xfs offline migrations
- Add tests

### CODE REVIEW

File: igvm/hypervisor.py [hunk 1]
```diff
@@ -828,6 +835,48 @@ def device_to_netcat(self, device, size, listener):
...
 
+    @contextmanager
+    def xfsrestore(self, vm: VM, transaction: Transaction = None):
+        device = self.get_volume_by_vm(vm).path()
+        port = self.netcat_port(device)
+        mount_dir = self.format_vm_storage(vm, transaction)
+
+        # Check xfsrestore exists, since problem is hidden with
+        # 'nohup', '&' and 'pty=False'
+        self.run('xfsrestore --help', silent=True)
```
Comment (TuxPowered42) [inline igvm/hypervisor.py:894]: We are never checking for any other programs as we expect them to be deployed by Puppet. Why do we need to check for this one? And if we go with the way of checking, we should add even more checks, preferrably in another PR, there was even a ticket somewhere to check if it is possible to SSH to a VM before igvm starts, that seems a good idea to implemment too.

File: igvm/hypervisor.py [hunk 2]
```diff
@@ -828,6 +835,48 @@ def device_to_netcat(self, device, size, listener):
             .format(device, size, listener[0], listener[1])
         )
 
+    @contextmanager
+    def xfsrestore(self, vm: VM, transaction: Transaction = None):
+        device = self.get_volume_by_vm(vm).path()
+        port = self.netcat_port(device)
+        mount_dir = self.format_vm_storage(vm, transaction)
+
+        # Check xfsrestore exists, since problem is hidden with
```
Comment (TuxPowered42) [inline igvm/hypervisor.py:892]: "if ... exists"

File: igvm/hypervisor.py [hunk 3]
```diff
@@ -828,6 +835,48 @@ def device_to_netcat(self, device, size, listener):
...
+        )
+        try:
+            yield self.fqdn, port
+        except BaseException:
+            self.kill_netcat(port)
+            raise
+
+    def xfsdump(self, vm: VM, listener, transaction: Transaction = None):
+        mount_dir = self.mount_vm_storage(vm, transaction)
+        # Check xfsdump exists, since problem is hidden with pipe
```
Comment (TuxPowered42) [inline igvm/hypervisor.py:914]: "if ... exists"

**Review by TuxPowered42 — CHANGES_REQUESTED** [review]

File: igvm/hypervisor.py [hunk 4]
```diff
@@ -567,37 +566,45 @@ def migrate_vm(
                 .format(offline_transport)
             )
         disk_size = disk_size if 0 < disk_size else 0
-        if offline_transport != 'xfs' and 0 < disk_size:
+        if not (offline and offline_transport == 'xfs') and 0 < disk_size:
```
[the remark this answers is not in the document]
Comment (Felixoid) [reply igvm/hypervisor.py:569]: I'm trying to find an article where I've read about the positive side of using `<` always.

If shortly, because it uses the "natural" x axis `------->`, it allows decreasing the brain load to parse the conditions.

I'm doing my best to find the article. =\

File: igvm/cli.py [hunk 5]
```diff
@@ -165,8 +165,10 @@ def parse_args():
     subparser.add_argument(
         '--offline-transport',
         default='drbd',
+        choices=('drbd', 'netcat', 'xfs'),
         help=(
-            'Specify drbd (default) or netcat transport to migrate disk image'
+            'Specify drbd (default), netcat or xfs transport to migrate '
```
[the remark this answers is not in the document]
Comment (Felixoid) [reply igvm/cli.py:170]: I'll add the formatter, because I like the idea of displaying defaults.

File: igvm/hypervisor.py [hunk 6]
```diff
@@ -606,35 +621,15 @@ def migrate_vm(
...
-                            transaction=transaction,
-                        )
-
+                self.migrate_wait_shutdown(vm, no_shutdown, transaction)
                 vm_disk_path = target_hypervisor.get_volume_by_vm(vm).path()
                 with target_hypervisor.netcat_to_device(vm_disk_path) as args:
                     self.device_to_netcat(
                         self.get_volume_by_vm(vm).path(),
-                        vm.dataset_obj['disk_size_gib'] * 1024 ** 3,
+                        vm.dataset_obj['disk_size_gib'] * (1 << 30),
```
[the remark this answers is not in the document]
Comment (Felixoid) [reply igvm/hypervisor.py:680]: at least that's how it's used everywhere in our puppet. From the preferences, I like more the bit notation.

File: igvm/cli.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Felixoid) [reply igvm/cli.py:170]: Changed my mind. It's out of the current scope since I'm not able to follow the logic of `IGVMArgumentParser` quickly.

File: igvm/hypervisor.py [hunk 7]
```diff
@@ -559,12 +559,24 @@ def vm_block_device_name(self):
...
         offline_transport: str, transaction: Transaction, no_shutdown: bool,
+        disk_size: int = 0,
     ):
         if offline_transport not in ['netcat', 'drbd', 'xfs']:
             raise StorageError(
                 'Unknown offline transport method {}!'
                 .format(offline_transport)
             )
+        disk_size = disk_size if 0 < disk_size else 0
+        if offline_transport != 'xfs' and 0 < disk_size:
```
[the remark this answers is not in the document]
Comment (Felixoid) [reply igvm/hypervisor.py]: Here I've found a hint why is it better https://flylib.com/books/en/2.823.1.155/1/, Search on the page by `Writing Numeric Expressions in Number-Line Order`

File: igvm/hypervisor.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Felixoid) [reply igvm/hypervisor.py:569]: Here I've found a hint why is it better https://flylib.com/books/en/2.823.1.155/1/, Search on the page by `Writing Numeric Expressions in Number-Line Order`

**Comment by Felixoid** [comment]
retest it

File: igvm/hypervisor.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (kofrezo) [reply igvm/hypervisor.py:680]: I agree with @TuxPowered42 this is way less readable please keep the current way of writing.

File: igvm/hypervisor.py [hunk 8]
```diff
@@ -802,18 +797,22 @@ def check_netcat(self, port):
     def kill_netcat(self, port):
         self.run('pkill -f "^/bin/nc.openbsd -l -p {}"'.format(port))
 
-    @contextmanager
-    def netcat_to_device(self, device):
+    def netcat_port(self, device):
```
Comment (kofrezo) [inline igvm/hypervisor.py:855]: I think it is a good idea to split this into a dedicated method :+1:. 

It might be also worth to make it a underscore method `_netcat_port` to hint users that this is a internal method that should not get used outside of the module.

File: igvm/hypervisor.py [hunk 9]
```diff
@@ -802,18 +797,22 @@ def check_netcat(self, port):
...
         dev_minor = self.run('stat -L -c "%T" {}'.format(device), silent=True)
         dev_minor = int(dev_minor, 16)
         port = 7000 + dev_minor
-
         self.check_netcat(port)
+        return port
+
+    @contextmanager
+    def netcat_to_device(self, device):
```
Comment (kofrezo) [inline igvm/hypervisor.py:863]: A doc string would be helpful here.

File: igvm/hypervisor.py [hunk 10]
```diff
@@ -555,9 +556,23 @@ def vm_block_device_name(self):
         the guest OS"""
         return 'vda1'
 
+    def migrate_wait_shutdown(
+        self, vm: VM, no_shutdown: bool, transaction: Transaction,
+    ):
```
Comment (kofrezo) [inline igvm/hypervisor.py:610]: This should method should be a parameter of `migrate_vm`.

File: igvm/hypervisor.py [hunk 11]
```diff
@@ -555,9 +556,23 @@ def vm_block_device_name(self):
...
+                vm.shutdown(
+                    check_vm_up_on_transaction=False,
+                    transaction=transaction,
+                )
+
     def migrate_vm(
-        self, vm, target_hypervisor, offline, offline_transport, transaction,
-        no_shutdown,
+        self, vm: VM, target_hypervisor: 'Hypervisor', offline: bool,
+        offline_transport: str, transaction: Transaction, no_shutdown: bool,
```
Comment (kofrezo) [inline igvm/hypervisor.py:624]: You can make adding type hints a dedicated commit the next time to make it easier for people to distinct style changes from actual changes.

File: igvm/cli.py [hunk 12]
```diff
@@ -185,6 +185,14 @@ def parse_args():
         action='store_true',
         help='Build or migrate VM only to a HV with the same environment of VM'
     )
+    subparser.add_argument(
+        '--disk-size',
+        dest='disk_size',
+        default=0,
```
[the remark this answers is not in the document]
Comment (kofrezo) [reply igvm/cli.py:191]: If it is a required argument to the command and and should not be possible to run it without specifying a disk size you can make it a positional argument.

File: igvm/commands.py [hunk 13]
```diff
@@ -391,6 +391,13 @@ def vm_migrate(vm_hostname=None, vm_object=None, hypervisor_hostname=None,
                     _vm.dataset_obj['datacenter_type'])
             )
 
+        # We have to check migration settings before searching for a HV,
+        # because the new disk size must be checked and set
+        current_size_gib = _vm.dataset_obj['disk_size_gib']
+        _vm.dataset_obj['disk_size_gib'] = _vm.hypervisor.migrate_disk_size(
+            _vm, offline, offline_transport, disk_size
+        )
```
Comment (kofrezo) [inline igvm/commands.py:399]: Does this also work for the algorithm to automatically choose a Hypervisor ?

File: igvm/hypervisor.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (kofrezo) [reply igvm/hypervisor.py:569]: I agree with @TuxPowered42. 

Even if your article has an interesting point in here we should stick to what we use more often in the whole project and not change the style in at some places in the project.

Please take this into account for further comparisons too.

File: igvm/hypervisor.py [hunk 14]
```diff
@@ -555,15 +556,77 @@ def vm_block_device_name(self):
...
+        if 0 < disk_size < allocated_space:
+            raise StorageError(
+                'disk_size is lower than allocated space: {} < {}!'
+                .format(disk_size, allocated_space)
+            )
+
+    def migrate_disk_size(
+        self, vm: VM, offline: bool, offline_transport: str,
+        disk_size: int = 0,
+    ) -> int:
```
Comment (kofrezo) [inline igvm/hypervisor.py:583]: I was confused in the first place when reading the method name because I thought it would actually "migrate" or change the disk size. Maybe we can find a better name for this by inverting like e.g. `disk_size_for_migration`

File: igvm/hypervisor.py [hunk 15]
```diff
@@ -555,15 +556,77 @@ def vm_block_device_name(self):
...
+        self.check_migrate_vm(vm, offline, offline_transport, disk_size)
+        disk_size = disk_size if 0 < disk_size else 0
+        return disk_size or vm.dataset_obj['disk_size_gib']
+
+    def migrate_change_disk_size(
+        self, vm: VM, offline: bool, offline_transport: str,
+        transaction, disk_size: int = 0,
+    ):
```
Comment (kofrezo) [inline igvm/hypervisor.py:591]: Same here. The name does not really say what the method is doing IMHO. Maybe we use something like `apply_new_disk_size` or `save_new_disk_size` and a doc string.

File: igvm/hypervisor.py [hunk 16]
```diff
@@ -555,15 +556,77 @@ def vm_block_device_name(self):
...
+                vm.dataset_obj.commit()
+            transaction.on_rollback('reset_disk_size', restore_size)
+
+
+    def migrate_wait_shutdown(
+        self, vm: VM, no_shutdown: bool, transaction: Transaction,
+    ):
+        vm.set_state('maintenance', transaction=transaction)
+        if vm.is_running():
+            if no_shutdown:
```
Comment (kofrezo) [inline igvm/hypervisor.py:613]: I was not expecting that - maybe `wait_for_shutdown` or `manual_shutdown` is a better name.

File: igvm/hypervisor.py [hunk 17]
```diff
@@ -555,15 +556,77 @@ def vm_block_device_name(self):
...
+                log.info('Please shut down the VM manually now')
+                vm.wait_for_running(running=False, timeout=86400)
```
Comment (kofrezo) [inline igvm/hypervisor.py:615]: I don't think it is realistic that we have to wait 24 hours to shutdown a VM. I suggest using a smaller timeout such as 5 minutes and check again if the VM was shut down and if not raise an Exception.

File: igvm/hypervisor.py [hunk 18]
```diff
@@ -826,8 +880,50 @@ def device_to_netcat(self, device, size, listener):
         self.run(
             'dd if={0} ibs=1048576 | pv -f -s {1} '
             '| /bin/nc.openbsd -q 1 {2} {3}'
-            .format(device, size, *listener)
+            .format(device, size, listener[0], listener[1])
+        )
+
+    @contextmanager
+    def xfsrestore(self, vm: VM, transaction: Transaction = None):
```
Comment (kofrezo) [inline igvm/hypervisor.py:887]: This a quite powerful and new method please add a doc string.

File: igvm/hypervisor.py [hunk 19]
```diff
@@ -826,8 +880,50 @@ def device_to_netcat(self, device, size, listener):
...
+            .format(port, mount_dir),
+            pty=False,  # Has to be here for background processes
+        )
+        try:
+            yield self.fqdn, port
+        except BaseException:
+            self.kill_netcat(port)
+            raise
+
+    def xfsdump(self, vm: VM, listener, transaction: Transaction = None):
```
Comment (kofrezo) [inline igvm/hypervisor.py:912]: Same here. Please add a doc string.

File: igvm/hypervisor.py (hunk 1 above)
Comment (kofrezo) [reply igvm/hypervisor.py:894]: I agree with @TuxPowered42 here. 

We have certain configuration requirements to a environment such as that some software is installed. This is part of the configuration management and should not be checked here. 

Adding such checks can be useful when having a transition from one OS to another (Stretch to Buster) in combination with conditionals but there is no need for this here.

File: igvm/hypervisor.py [hunk 20]
```diff
@@ -826,8 +880,50 @@ def device_to_netcat(self, device, size, listener):
...
+        # xfsrestore needs to output its logs, otherwise it fails
+        self.run(
+            'nohup /bin/nc.openbsd -l -p {0} '
+            '| xfsrestore -F -J - {1} 2>{1}.log 1>&2 &'
+            .format(port, mount_dir),
+            pty=False,  # Has to be here for background processes
+        )
+        try:
+            yield self.fqdn, port
+        except BaseException:
```
Comment (kofrezo) [inline igvm/hypervisor.py:908]: Why are you explicelty capturing BaseException here ?

File: tests/test_integration.py [hunk 21]
```diff
@@ -582,6 +582,45 @@ def test_offline_migration_drbd(self):
...
+        vm_migrate(
+            VM_HOSTNAME,
+            offline=True,
+            offline_transport='xfs',
+        )
+        self.check_vm_present()
+
+        disk_size_gib = self.vm_obj['disk_size_gib']
+
+        # Increase disk size during migration
```
Comment (kofrezo) [inline tests/test_integration.py:596]: This should be a dedicated test.

File: tests/test_integration.py [hunk 22]
```diff
@@ -582,6 +582,45 @@ def test_offline_migration_drbd(self):
...
+        vm_migrate(
+            VM_HOSTNAME,
+            offline=True,
+            offline_transport='xfs',
+            disk_size=disk_size_gib + 1,
+        )
+        self.check_vm_present()
+
+        # Decreasing disk size back during migration
```
Comment (kofrezo) [inline tests/test_integration.py:605]: This should be a dedicated test.

File: tests/test_integration.py [hunk 23]
```diff
@@ -582,6 +582,45 @@ def test_offline_migration_drbd(self):
...
+        vm_migrate(
+            VM_HOSTNAME,
+            offline=True,
+            offline_transport='xfs',
+            disk_size=disk_size_gib,
+        )
+        self.check_vm_present()
+
+        # Attempt to decrease disk size lower than allocated space
+        with self.assertRaises(IGVMError):
```
Comment (kofrezo) [inline tests/test_integration.py:615]: This should be a dedicated test.

File: tests/test_integration.py [hunk 24]
```diff
@@ -643,6 +682,21 @@ def test_rollback_drbd(self):
 
         self.check_vm_present()
 
+    def test_rollback_xfs(self):
+        obj = Query({'hostname': VM_HOSTNAME}, ['puppet_environment']).get()
```
Comment (kofrezo) [inline tests/test_integration.py:686]: You should be able to access it via `self.vm_obj` without doing an extra Query.

File: igvm/hypervisor.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Felixoid) [reply igvm/hypervisor.py:569]: More verbose description http://llewellynfalco.blogspot.com/2016/02/dont-use-greater-than-sign-in.html

**Review by kofrezo — CHANGES_REQUESTED** [review]
Thanks for adding this cool new feature, I am looking forward to use it in production. 

There are some things missing such as doc strings and some changes I disagree on. Please check them and lets discuss.

File: igvm/cli.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (Felixoid) [reply igvm/cli.py:191]: It is an optional argument, that works only together with `--offline --offline-transport xfs`.

If it's used without it:

```
Traceback (most recent call last):
  File "run.py", line 12, in <module>
    main()
  File "/home/mikhail.shiryaev/igvm/igvm/cli.py", line 440, in main
    args.pop('func')(**args)
  File "/home/mikhail.shiryaev/igvm/igvm/host.py", line 25, in decorator
    return fn(*args, **kwargs)
  File "/home/mikhail.shiryaev/igvm/igvm/commands.py", line 398, in vm_migrate
    _vm, offline, offline_transport, disk_size
  File "/home/mikhail.shiryaev/igvm/igvm/hypervisor.py", line 584, in migrate_disk_size
    self.check_migrate_vm(vm, offline, offline_transport, disk_size)
  File "/home/mikhail.shiryaev/igvm/igvm/hypervisor.py", line 571, in check_migrate_vm
    'disk_size can be applied only with offline transport xfs!'
igvm.exceptions.StorageError: disk_size can be applied only with offline transport xfs!
```

File: igvm/hypervisor.py (hunk 17 above)
Comment (Felixoid) [reply igvm/hypervisor.py:615]: This method is an unchanged snippet from [1](https://github.com/innogames/igvm/pull/176/commits/5fff38a34b71643951dc72192c65625535f06240#diff-d587443efd90ee0cdc812d478981451c4b4a15000f584b1fc06ea3256d154fddL609-L619) and [2](https://github.com/innogames/igvm/pull/176/commits/5fff38a34b71643951dc72192c65625535f06240#diff-d587443efd90ee0cdc812d478981451c4b4a15000f584b1fc06ea3256d154fddL622-L632)

File: igvm/hypervisor.py (hunk 10 above)
Comment (Felixoid) [reply igvm/hypervisor.py:610]: Don't get it. What actually should be a parameter? All of the arguments of this method are parameters of `migrate_vm`

File: igvm/hypervisor.py (hunk 16 above)
Comment (Felixoid) [reply igvm/hypervisor.py:613]: The parameter is the CLI argument, so `no_shutdown` is coming from the parser. I'll rename the method though.

File: igvm/hypervisor.py (hunk 20 above)
Comment (Felixoid) [reply igvm/hypervisor.py:908]: This is the same usage as in `netcat_to_device`. I suspect this is done to keep flake8 silent about `cathing everything` error.

File: tests/test_integration.py (hunk 24 above)
Comment (Felixoid) [reply tests/test_integration.py:686]: True, but it's done in the same way for every rollback test. Should I change it for everything else too?

File: igvm/cli.py [hunk 25]
```diff
@@ -185,6 +185,14 @@ def parse_args():
...
+        type=int,
+        help='When --offline-transport=xfs is used, this parameter may specify'
```
[the remark this answers is not in the document]
Comment (TuxPowered42) [reply igvm/cli.py:193]: It is an optional argument, once it is specified it does things always. Saying "may" means that it could still be ignored depending on other conditions.

File: igvm/hypervisor.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (TuxPowered42) [reply igvm/hypervisor.py]: Such notation makes perfect sense for testing ranges. But for single numbers…? I dunno. @kofrezo is thre a PEP for that? I've checked PEP-8 and it's not specified there, however all examples do `VAR > VAL`, not the opposite.

File: igvm/cli.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (TuxPowered42) [reply igvm/cli.py:191]: `0` is a valid (not for this particular option but I'm talking coding in general) value which user might enter. `None` means the option was not given at all. Those are separate things and using Magic Numbers is not necessary in languages like Python, where we can easily check if things exist at all. 

Back to the other piece of code, where it is used:
```
disk_size = disk_size if 0 < disk_size else 0
if not (offline and offline_transport == 'xfs') and 0 < disk_size:
```

Could just be:
```
if disk_size and disk_size < 1:
    raise StorageError('New disk size must be at least 1GiB!')
if disk_size and not (offline and offline_transport == 'xfs'):
    raise StorageError('Disk can only be resized with offline migration and --offline_transport=xfs!')
```

File: tests/test_integration.py (hunk 24 above)
Comment (TuxPowered42) [reply tests/test_integration.py:686]: If every other test queries a new object, maybe it is necessary for something? I vaguely remember there was some magic happening here.

File: igvm/hypervisor.py (hunk 17 above)
Comment (TuxPowered42) [reply igvm/hypervisor.py:615]: This function was designed for migration of Graphite hosts and they needed a very long time to shut down. Let's just keep it as it is.

File: tests/test_integration.py (hunk 21 above)
Comment (Felixoid) [reply tests/test_integration.py:597]: Done

File: tests/test_integration.py (hunk 22 above)
Comment (Felixoid) [reply tests/test_integration.py:606]: I would leave it as one test for disk changing since first it increases the disk and then decreases to the original size.

File: tests/test_integration.py (hunk 23 above)
Comment (Felixoid) [reply tests/test_integration.py:617]: done

File: tests/test_integration.py (hunk 24 above)
Comment (Felixoid) [reply tests/test_integration.py:706]: I'd like to make it consistent with other tests for now. If you want, I could leave a TODO comment for future

File: igvm/hypervisor.py (hunk 15 above)
Comment (Felixoid) [reply igvm/hypervisor.py:600]: Changed. Looks better?

File: igvm/hypervisor.py (hunk 16 above)
Comment (Felixoid) [reply igvm/hypervisor.py:621]: Renamed to `wait_for_shutdown`

File: igvm/hypervisor.py (hunk 14 above)
Comment (Felixoid) [reply igvm/hypervisor.py:589]: Renamed to `vm_new_disk_size`

**Comment by Felixoid** [comment]
I will squash commits tomorrow to ease the review since now it doesn't look pretty. So only 5 commits will be in place

File: igvm/hypervisor.py (hunk 15 above)
Comment (kofrezo) [reply igvm/hypervisor.py:600]: Why not `apply_new_vm_disk_size` - this is more close to our every day English than `apply_vm_new_disk_size` ?

File: igvm/hypervisor.py (hunk 10 above)
Comment (kofrezo) [reply igvm/hypervisor.py:618]: My bad I missed how it is used in migrate_vm. This is fine.

You could still add a doc string and consider making it a underscore method.

File: igvm/hypervisor.py (hunk 16 above)
Comment (kofrezo) [reply igvm/hypervisor.py:621]: We should not allow changing the disk size without offline in the first place then. Similar to this: 
-  https://github.com/innogames/igvm/blob/master/igvm/commands.py#L418

Making the user think he can migrate the VM and change the disk size online and then ask him to shutdown the VM is counter intuitive and does not fit our already existing handling of similar cases. We are better of telling him right away that this is not possible.

File: igvm/hypervisor.py (hunk 17 above)
Comment (kofrezo) [reply igvm/hypervisor.py:623]: Okay but lets at least tell the user than how long we will wait.

File: igvm/hypervisor.py (hunk 20 above)
Comment (kofrezo) [reply igvm/hypervisor.py:958]: Let me rephrase the questions: Why do you expect yielding two primitive data types (a string and a integer) to raise a BaseException ?

If it is because of the method `_netcat_port` raising it we should consider using Transaction.

File: tests/test_integration.py (hunk 22 above)
Comment (kofrezo) [reply tests/test_integration.py:606]: This are two scenarios for the user. One might increase a disk size with IGVM but never decrease it and vice versa.

Additionally having many small test methods allows us to run tests in parallel and will speed up the test pipeline.

File: tests/test_integration.py (hunk 24 above)
Comment (kofrezo) [reply tests/test_integration.py:706]: Fair enough. Lets just make a todo.

**Review by kofrezo — CHANGES_REQUESTED** [review]

File: igvm/hypervisor.py (hunk 15 above)
Comment (Felixoid) [reply igvm/hypervisor.py:600]: I wanted to make it consistent with `vm_new_disk_size` -> `apply_vm_new_disk_size`. First vm indicates the scope.

Maybe, `_vm_apply_new_disk_size` would sound better. The method looks private to me.

File: igvm/hypervisor.py (hunk 20 above)
Comment (Felixoid) [reply igvm/hypervisor.py:958]: From my side, I've just reused the same snippet from https://github.com/innogames/igvm/commit/7ed936d4#diff-d587443efd90ee0cdc812d478981451c4b4a15000f584b1fc06ea3256d154fddR743

Maybe, the contextmanager API is used wrong in the first place, but as far as I know, @TuxPowered42 would like to get rid of Transaction and to move to the context lib.

File: igvm/hypervisor.py (hunk 16 above)
Comment (Felixoid) [reply igvm/hypervisor.py:621]: I guess the last comment is related to some other discussion.

Nevertheless, it looks more reasonable than to just replace https://github.com/innogames/igvm/blob/master/igvm/commands.py#L418 with `check_migrate_parameters`

File: tests/test_integration.py (hunk 22 above)
Comment (Felixoid) [reply tests/test_integration.py:606]: I see the point. But the parallel execution has the overhead of VM deployment, and to be sure that the decrease operation will work, I have to increase the disk in advance.

If you insist, it's possible to apply `vm_set_disk_size_gib` before xfs migration with disk size decrease

File: igvm/hypervisor.py (hunk 20 above)
Comment (Felixoid) [reply igvm/hypervisor.py:958]: Here's the answer, why `BaseException` is caught https://docs.python.org/3.5/library/contextlib.html#contextlib.contextmanager

> If an unhandled exception occurs in the block, it is reraised inside the generator at the point where the yield occurred.

So, if anything happens at https://github.com/innogames/igvm/pull/176/files#diff-d587443efd90ee0cdc812d478981451c4b4a15000f584b1fc06ea3256d154fddR694, the `@contextmanager` will clean up the netcat server on the target hypervisor.

File: igvm/hypervisor.py (hunk 16 above)
Comment (Felixoid) [reply igvm/hypervisor.py:621]: And `check_migrate_parameters` are already included here https://github.com/innogames/igvm/pull/176/commits/0371e62cd780a9044b36498ac7971049410228fb#diff-9299a928e2b4f7d2b1d51a5dcdd44a442d587f35480a7f2dfe8349022180e2cdR397, so the migration won't start on the wrong parameters

File: igvm/hypervisor.py (hunk 17 above)
Comment (Felixoid) [reply igvm/hypervisor.py:642]: Added to argparse help

File: tests/test_integration.py (hunk 22 above)
Comment (Felixoid) [reply tests/test_integration.py:626]: Split the test into two different

File: tests/test_integration.py (hunk 24 above)
Comment (Felixoid) [reply tests/test_integration.py:733]: TODO added as a dedicated commit

File: igvm/hypervisor.py (hunk 15 above)
Comment (kofrezo) [reply igvm/hypervisor.py:602]: Yes I think that is much better.

File: igvm/hypervisor.py (hunk 17 above)
Comment (kofrezo) [reply igvm/hypervisor.py:642]: Fair enough.

File: igvm/hypervisor.py (hunk 20 above)
Comment (kofrezo) [reply igvm/hypervisor.py:983]: Ah makes sense. Thanks for explaining.

**Review by kofrezo — APPROVED** [review]

**Comment by kofrezo** [comment]
We still need to figure out a way to make the newly added tests pass always because the disk resize sometimes fails (for a unknown reason) but for the status quo of code you have my green light.

**Comment by kofrezo** [comment]
> We still need to figure out a way to make the newly added tests pass always because the disk resize sometimes fails (for a unknown reason) but for the status quo of code you have my green light.

Already tackled by:

- https://github.com/innogames/igvm/pull/176/commits/21c311efed2158b4b99241e3b85dac8ace016c92
- https://github.com/innogames/igvm/pull/176/commits/116db75c36702e430fcfe058a69b37f96f14516c

File: igvm/hypervisor.py [hunk 26]
```diff
@@ -800,34 +885,143 @@ def check_netcat(self, port):
...
+    def _xfsrestore_log_name(self, vm: VM) -> str:
+        device_name = self.get_volume_by_vm(vm).name()
+        return '/tmp/xfsrestore-{}.log'.format(device_name)
+
+    def check_xfsrestore_log(self, vm: VM):
+        """
+        Search for WARNING in the xfsrestore log file.
+        Raises exception if found
+        """
+        self.run('cat {}'.format(self._xfsrestore_log_name(vm)))
```
Comment (TuxPowered42) [inline igvm/hypervisor.py:954]: Why first we cat the file and then grep it?

Comment (Felixoid) [reply igvm/hypervisor.py:954]: To have the log output in the migration log. Only for this.

Comment (TuxPowered42) [reply igvm/hypervisor.py:954]: There is no migration log. If you mean dumping the output to  console, that's fine.

**Review by TuxPowered42 — APPROVED** [review]

**Comment by Felixoid** [comment]
Thank you for your review, gentlemen.

**Comment by Felixoid** [comment]
@ig-sysadmin-jenkins retest this please

**Comment by Felixoid** [comment]
@ig-sysadmin-jenkins retest this please
